addons/website_sale_referral/controllers/referral.py

```diff
@@ -0,0 +1,109 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from odoo.http import Controller, route, request
+import uuid
+import json
+
+from werkzeug.exceptions import Forbidden, BadRequest
+
+
+class Referral(Controller):
+
+    @route(['/referral'], type='http', auth='public', website=True)
+    def referral_unauth(self, force=False, **kwargs):
+        if not force and not request.website.is_public_user():
+            if request.env.user.partner_id.referral_tracking_id:
+                token = request.env.user.partner_id.referral_tracking_id.token
+                return request.redirect('/referral/' + token)
+            else:
+                return self.referral_register(request.env.user.partner_id.email)
+        else:
+            return request.render('website_sale_referral.referral_controller_template_register')
+
+    @route(['/referral/register'], type='http', auth='public', method='POST', website=True)
+    def referral_register(self, referrer_email, token=None, **post):
+        if token:
+            referral_tracking = request.env['referral.tracking'].sudo().search([('token', '=', token)], limit=1)
+            if referral_tracking:
+                if referral_tracking.referrer_email != referrer_email:
+                    raise Forbidden()  # Mismatch between email and token
+                return request.redirect('/referral/' + referral_tracking.token)
+        else:
+            token = uuid.uuid4().hex[:-1]  # to avoid conflict with internal token
+
+        utm_name = ('%s-%s') % (referrer_email, str(uuid.uuid4())[:6])
```

Comment:
We should still change the utm_name generation using a uuid4 only to avoid exposing the email address on the shared links. 

@bfr-o Could you handle this ? Thanks ;)

Reply:
@tivisse this is done on the branch 13.0-website-referral-bfr, maybe we should open a new PR with 13.0 as a target or should I apply the changes on this branch too ?

Reply:
Oh, my bad. I was on the wrong PR. 

No it's ok like this ;)